code/game/machinery/doors/door.dm

```diff
@@ -67,8 +80,45 @@
 	health = maxhealth
 
 	update_nearby_tiles(need_rebuild=1)
+	if (hashatch)
+		setup_hatch()
 	return
 
+/obj/machinery/door/proc/setup_hatch()
+
+	if (overlays != null)
+		hatch_image = image('icons/obj/doors/hatches.dmi', src, hatchstyle, closed_layer+0.1)
+		hatch_image.color = hatch_colour
+		hatch_image.pixel_x = hatch_offset_x
+		hatch_image.pixel_y = hatch_offset_y
+
+		overlays += hatch_image
+		update_icon()
+	else
+		spawn(10)
+			setup_hatch()
+
+/obj/machinery/door/proc/open_hatch(var/atom/mover = null)
+	if (!hatchstate)
+		hatchstate = 1
+		update_icon()
+		playsound(src.loc, hatch_open_sound, 40, 1, -1)
+	hatchclosetime = world.time + 29
+
+	if (istype(mover, /mob/living/silicon))
+		var/mob/living/silicon/S = mover
+		S.under_door()
+	spawn(30)

```

Comment:
This was a conscious choice because doors and airlocks have their own process() each with different code that doesn't inherit. Also process procs every 2.1 seconds so its impossible to get a close time of 3 seconds except by sheer luck.

Using spawn alone causes the hatches to desync a bit, and makes for some screwy behaviour if the same hatch is used a few times in a short span.

I fail to see what's dangerous about this approach. The various spawn commands each just push forward the time, so the door will close exactly when desired, three seconds after it was last used